social-network-api/src/main/java/com/socialnetwork/api/controller/PostController.java

```diff
@@ -1,90 +1,137 @@
-  //  }
-
-  //  @DeleteMapping()
-  //  public ResponseEntity<?> deletePost(@RequestBody PostDto postDto) {
-  //    Post post = convertToPost(postDto);
-  //    if (!postService.existsById(post.getPostId())) {
-  //      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new Response(POST_NOT_FOUND));
-  //    }
-  //    postService.delete(post);
-  //    return ResponseEntity.ok(new Response("Post was deleted"));
-  //  }
-
-  private Post convertToPost(PostDto postDto) {
-    return modelMapper.map(postDto, Post.class);
+  @PostMapping("/bookmark")
+  public PostDto.Response.Default
+    saveBookmark(@RequestBody PostDto.Request.Action postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Post post = postService.getReferenceById(postDto.getPost().getId());
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
+
+    if (bookmarkService.existsByUserAndPost(user, post)) {
+      bookmarkService.delete(user, post);
+    } else {
+      bookmarkService.save(user, post);
+    }
+
+    return convertToPostDto(post);
+  }
+
+  @PostMapping("/like")
+  public PostDto.Response.Default
+    saveLike(@RequestBody PostDto.Request.Action postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Post post = postService.getReferenceById(postDto.getPost().getId());
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
+
+    if (likeService.existsByUserAndPost(user, post)) {
+      likeService.delete(user, post);
+    } else {
+      likeService.save(user, post);
+    }
+
+    return convertToPostDto(post);
+  }
+
+  private Post convertToPost(PostDto.Request.Created postDto, User user, Integer originalPostId)
+          throws NoPostWithSuchIdException {
+    Post post = modelMapper.map(postDto, Post.class);
+    post.setAuthor(user);
+
+    if (originalPostId != null && postService.existsById(originalPostId)) {
+      post.setOriginalPost(postService.getReferenceById(originalPostId));
+    }
+
+    return post;
+  }
+
+
+  private PostDto.Response.Default convertToPostDto(Post post) {
+    PostDto.Response.Default postDto = modelMapper.map(post, PostDto.Response.Default.class);
+    postDto.setLikesNumber(likeService.countPostLikes(post));
+    postDto.setBookmarksNumber(bookmarkService.countPostBookmarks(post));
+
+    if (postDto.getOriginalPost() != null) {
+      Post originalPost = postService.findById(postDto.getOriginalPost().getId()).get();
+      postDto.getOriginalPost().setLikesNumber(likeService.countPostLikes(originalPost));
```

Comment:
Let frontend take the number of likes from the original post if it`s a retweet. There is no need for the additional counter for them in the retweet.